### PR TITLE
Document property classes

### PR DESCRIPTION
**Pull request by enbrain**

### CODE REVIEW

File: mappings/net/minecraft/state/property/BooleanProperty.mapping [hunk 1]
```diff
@@ -1,6 +1,13 @@
 CLASS net/minecraft/class_2746 net/minecraft/state/property/BooleanProperty
+	COMMENT Represents a property that can take boolean values.
```
Comment (liach) [inline mappings/net/minecraft/state/property/BooleanProperty.mapping]: A property not only takes values, but it only has a limited set of allowed values.
```suggestion
	COMMENT Represents a property that has boolean values.
```

File: mappings/net/minecraft/state/property/DirectionProperty.mapping [hunk 2]
```diff
@@ -1,24 +1,42 @@
...
 		ARG 2 values
 	METHOD method_11843 of (Ljava/lang/String;Ljava/util/Collection;)Lnet/minecraft/class_2753;
-		COMMENT Creates a direction property which only supports specific values
+		COMMENT Creates a direction property with the given values.
+		COMMENT
+		COMMENT @see #of(String)
 		ARG 0 name
-			COMMENT the name of this property
+			COMMENT the name of the property; required to match the pattern {@link
+			COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN}
```
Comment (liach) [inline mappings/net/minecraft/state/property/DirectionProperty.mapping]:
```suggestion
			COMMENT the name of the property; required to match the {@linkplain
			COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN valid name pattern}
```

File: mappings/net/minecraft/state/property/BooleanProperty.mapping [hunk 3]
```diff
@@ -1,6 +1,13 @@
...
+	COMMENT <p>See {@link net.minecraft.state.property.Properties} for example
+	COMMENT usages.
 	FIELD field_12575 values Lcom/google/common/collect/ImmutableSet;
 	METHOD <init> (Ljava/lang/String;)V
 		ARG 1 name
 	METHOD method_11825 of (Ljava/lang/String;)Lnet/minecraft/class_2746;
+		COMMENT Creates a boolean property.
 		ARG 0 name
+			COMMENT the name of the property; required to match the pattern {@link
+			COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN}
```
Comment (liach) [inline mappings/net/minecraft/state/property/BooleanProperty.mapping]: Similar suggestions on the valid name pattern link display

File: mappings/net/minecraft/state/property/DirectionProperty.mapping [hunk 4]
```diff
@@ -1,24 +1,42 @@
 CLASS net/minecraft/class_2753 net/minecraft/state/property/DirectionProperty
+	COMMENT Represents a property that can take direction values.
```
Comment (liach) [inline mappings/net/minecraft/state/property/DirectionProperty.mapping]: Same remark on "can take", "has" is better imo

File: mappings/net/minecraft/state/property/DirectionProperty.mapping [hunk 5]
```diff
@@ -1,24 +1,42 @@
...
+	COMMENT <p>See {@link net.minecraft.state.property.Properties} for example
+	COMMENT usages.
 	METHOD <init> (Ljava/lang/String;Ljava/util/Collection;)V
 		ARG 1 name
 		ARG 2 values
 	METHOD method_11843 of (Ljava/lang/String;Ljava/util/Collection;)Lnet/minecraft/class_2753;
-		COMMENT Creates a direction property which only supports specific values
+		COMMENT Creates a direction property with the given values.
+		COMMENT
+		COMMENT @see #of(String)
```
Comment (liach) [inline mappings/net/minecraft/state/property/DirectionProperty.mapping:12]: Consider documenting the exception (at least the type) thrown if the name/allowed values aren't valid

File: mappings/net/minecraft/state/property/DirectionProperty.mapping [hunk 6]
```diff
@@ -1,24 +1,42 @@
...
-			COMMENT the name of this property
+			COMMENT the name of the property; required to match the pattern {@link
+			COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN}
 		ARG 1 values
-			COMMENT the values this property can have
+			COMMENT the values the property can take; required to have 2 or more values
 	METHOD method_11844 of (Ljava/lang/String;Ljava/util/function/Predicate;)Lnet/minecraft/class_2753;
-		COMMENT Creates a direction property.
+		COMMENT Creates a direction property with the values allowed by the given
+		COMMENT filter.
```
Comment (liach) [inline mappings/net/minecraft/state/property/DirectionProperty.mapping]:
```suggestion
		COMMENT filter out of all 6 directions.
```

File: mappings/net/minecraft/state/property/EnumProperty.mapping [hunk 7]
```diff
@@ -1,33 +1,84 @@
...
 		ARG 3 values
 	METHOD method_11847 of (Ljava/lang/String;Ljava/lang/Class;Ljava/util/Collection;)Lnet/minecraft/class_2754;
-		COMMENT Creates an enum property.
+		COMMENT Creates an enum property with the given values.
+		COMMENT
+		COMMENT <p>The enum class is required to have 2 or more values. The enum class
+		COMMENT is required to provide a name for each value by overriding {@link
+		COMMENT StringIdentifiable#asString()}. The names of the values are required to
+		COMMENT match the pattern {@link
+		COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN}.
```
Comment (liach) [inline mappings/net/minecraft/state/property/EnumProperty.mapping]: same name pattern link remark

File: mappings/net/minecraft/state/property/EnumProperty.mapping [hunk 8]
```diff
@@ -1,33 +1,84 @@
...
+		COMMENT @throws IllegalArgumentException if multiple values have the same name
+		COMMENT
+		COMMENT @see #of(String, Class)
 		ARG 0 name
-			COMMENT the name of this property
+			COMMENT the name of the property; required to match the pattern {@link
+			COMMENT net.minecraft.state.StateManager#VALID_NAME_PATTERN}
 		ARG 1 type
-			COMMENT the type this property contains
+			COMMENT the type of the values the property can take
```
Comment (liach) [inline mappings/net/minecraft/state/property/EnumProperty.mapping]: I prefer the old "contain" doc, as a property isn't really an infinite taker like a predicate can be, but has discrete values and is finite. Same as the "has" vs "can take" remark.

File: mappings/net/minecraft/state/property/EnumProperty.mapping [hunk 9]
```diff
@@ -1,33 +1,84 @@
...
 		ARG 2 values
-			COMMENT the values this property could contain
+			COMMENT the values the property can take; required to have 2 or more values
 	METHOD method_11848 of (Ljava/lang/String;Ljava/lang/Class;Ljava/util/function/Predicate;)Lnet/minecraft/class_2754;
-		COMMENT Creates an enum property.
+		COMMENT Creates an enum property with the values allowed by the given filter.
+		COMMENT
+		COMMENT <p>The enum class is required to have 2 or more values. The enum class
+		COMMENT is required to provide a name for each value by overriding {@link
+		COMMENT StringIdentifiable#asString()}. The names of the values are required to
```
Comment (liach) [inline mappings/net/minecraft/state/property/EnumProperty.mapping]: If we duplicate this info, can we move it to the class doc and add an anchor like `<p id="anchor">` and add a `<a href="#anchor">link</a>` to reduce duplicate and avoid having out-of-date info?

File: mappings/net/minecraft/state/property/IntProperty.mapping [hunk 10]
```diff
@@ -7,11 +11,14 @@ CLASS net/minecraft/class_2758 net/minecraft/state/property/IntProperty
 	METHOD method_11867 of (Ljava/lang/String;II)Lnet/minecraft/class_2758;
 		COMMENT Creates an integer property.
 		COMMENT
-		COMMENT <p>{@code min} must be non-negative and {@code max} must be greater than {@code min}.
+		COMMENT <p>Note that this method takes {@code O(max - min)} time as it computes
```
Comment (liach) [inline mappings/net/minecraft/state/property/IntProperty.mapping]: Should be `O(n)` as it's linear. Imo just mentioning that it computes all possible values suffices.

File: mappings/net/minecraft/state/property/IntProperty.mapping [hunk 11]
```diff
@@ -7,11 +11,14 @@ CLASS net/minecraft/class_2758 net/minecraft/state/property/IntProperty
...
+		COMMENT all possible values during the instantiation.
 		COMMENT
-		COMMENT <p>Note that this method takes O({@code max} - {@code min}) time as it computes all possible values during instantiation.
+		COMMENT @throws IllegalArgumentException if {@code min < 0}
+		COMMENT @throws IllegalArgumentException if {@code max <= min}
```
Comment (liach) [inline mappings/net/minecraft/state/property/IntProperty.mapping]: Can we merge these two clauses?

File: mappings/net/minecraft/state/property/Property.mapping [hunk 12]
```diff
@@ -11,13 +11,16 @@ CLASS net/minecraft/class_2769 net/minecraft/state/property/Property
 		ARG 1 o
 	METHOD method_11799 computeHashCode ()I
 	METHOD method_11898 getValues ()Ljava/util/Collection;
-		COMMENT Returns all possible values the property can take.
+		COMMENT Returns all possible values this property can take.
```
Comment (liach) [inline mappings/net/minecraft/state/property/Property.mapping]:
```suggestion
		COMMENT Returns all possible values of this property.
```

**Review by liach — COMMENTED** [review]
Imo we should probably have some overview documentation on the properties, such as a meta package-info java asking people to look at `Property` to understand the API, a few impls for utilities, and `Properties` for examples.

Similarly, a brief description on `Property` class itself would facilitate understanding property as a dimension for a state (some more intuitive analogy needed) and states' keys are cartesian products of all the possible property values, but these info probably belong to the state manager class.

**Review by liach — APPROVED** [review]

**Review by Player3324 — APPROVED** [review]
Looks good, thanks! I'll merge it now to catch the next snapshot.

**Pull request merged**